### ISSUE TITLE
Annotation Type mismatch: cannot convert from Class<...> to Class<? extends AbstractDeclarativeValidator>

### ISSUE THREAD

**Issue by rubenporras** [eclipse-jdt/eclipse.jdt.core#875]
After upgrading to Eclipse 2023-03, on usages of the following Xtext annotation (org.eclipse.xtext.validation.ComposedChecks)
```
@Retention(RetentionPolicy.RUNTIME)
@Target({ ElementType.TYPE})
public @interface ComposedChecks {

	Class<? extends AbstractDeclarativeValidator>[] validators();

}
```

I get the error "Type mismatch: cannot convert from Class<AvqScriptJavaValidator> to Class<? extends AbstractDeclarativeValidator>". However, the type `AvqScriptJavaValidator` extends `AbstractDeclarativeValidator`, see the screenshot:

![image](https://user-images.githubusercontent.com/43636626/226542945-ba126ba2-6d8e-473c-a77a-0f815dcec2da.png)

**Comment by rubenporras** [comment]
As additional information, the code was compiling without errors in Eclipse 2022-12

**Comment by iloveeclipse** [comment]
Please try to provide a self containing reproducer project.

**Comment by srikanth-sankaran** [comment]
Do you perhaps have annotation based null analysis turned on ? I am wondering if this is a duplicate of https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1007

**Comment by rubenporras** [comment]
Yes, I do have them enabled. I will disable them and try again.

**Comment by rubenporras** [comment]
Hi @srikanth-sankaran , I can confirm that disabling the JDT annotation base analysis by using 
`org.eclipse.jdt.core.compiler.annotation.nullanalysis=disabled`
fixes the problem.

Also, if instead of 
```
@CompilerCatalog(compilers = {XCheck.class})
public class AvqBaseCheckValidator extends AcfCheckValidator {

  @Override
  protected List<EPackage> getEPackages() {
    return null;
  }
}
```
I just have

```
@CompilerCatalog(compilers = {XCheck.class})
public class AvqBaseCheckValidator {

  protected List<EPackage> getEPackages() {
    return null;
  }
}
```

then the error marker also disappear. Do you think you still need a self contained reproducer?

**Comment by rubenporras** [comment]
@srikanth-sankaran , also removing the brackets as commented in ttps://github.com/eclipse-jdt/eclipse.jdt.core/issues/1007#issuecomment-1534834099 removes the marker. Just to make it clear, the following has also no marker:

```
@CompilerCatalog(compilers = XCheck.class)
public class AvqBaseCheckValidator extends AcfCheckValidator {

  @Override
  protected List<EPackage> getEPackages() {
    return null;
  }
}
```

**Comment by iloveeclipse** [comment]
> Do you think you still need a self contained reproducer?

We have standalone reproducer in https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1007 already. Thanks for confirmation, I will close this one as duplicate of https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1007.

**Comment by stephan-herrmann** [comment]
@rubenporras while fixing #1007 I'd like to double check that your issue is fixed as well. 

Where is the error reported, against the declaration of `AvqBaseCheckValidator`, against ` AvqScriptJavaValidator` or at a usage site? If the latter please show that code, too (two lines should suffice). I also need to see a minimal declaration of `AvqBaseCheckValidator` (no code): super types, annotations if any.

How exactly should `XCheck` be found, is there an import, or should it be found implicitly?

I guess that there must be some connection between `XCheck` and `AcfCheckValidator`. Perhaps the former is a member of the latter?

What is the type of `CompilerCatalog#compilers`? Plain `Class<?>` or does it declare some type bound?

**Comment by rubenporras** [comment]
Hi @stephan-herrmann , I did not finish to create the reproducer before the ticket was marked as duplicate, but I only needed to upload the files. The reproducer I had was:

```
package com.avaloq.tools.dsl.avqbase.validation;

import java.util.Map;

import org.eclipse.emf.common.util.DiagnosticChain;
import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.EObject;

import com.avaloq.tools.ddk.check.runtime.issue.ICheckValidatorImpl;
import com.avaloq.tools.ddk.check.runtime.validation.DefaultCheckValidator;


public class AcfCheckValidator extends DefaultCheckValidator {

  @Override
  protected Iterable<? extends ICheckValidatorImpl> internalCollectValidators(final String language, final EObject object) { // NOPMD

    return null;
  }

  @Override
  protected boolean internalValidate(final EClass eClass, final EObject eObject, final DiagnosticChain diagnostics, final Map<Object, Object> context) {
    return true;
  }

}
```

```
package com.avaloq.tools.dsl.avqbase.validation;

import java.util.List;

import org.eclipse.emf.ecore.EPackage;

@CompilerCatalog(compilers = {XCheck.class})
public class AvqBaseCheckValidator extends AcfCheckValidator {

  @Override
  protected List<EPackage> getEPackages() {
    return null;
  }
}

```
```
package com.avaloq.tools.dsl.avqbase.validation;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import com.avaloq.tools.ddk.check.runtime.issue.ICheckValidatorImpl;


/**
 * Annotation which allows compiler check catalogs to be defined in a language's
 * run-time plug-in.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.TYPE})
public @interface CompilerCatalog {
  Class<? extends ICheckValidatorImpl>[] compilers();
}

```
```
package com.avaloq.tools.dsl.avqbase.validation;

import com.avaloq.tools.ddk.check.runtime.issue.DefaultCheckImpl;
import com.google.common.collect.ImmutableMap;


public class XCheck extends DefaultCheckImpl {

  @Override
  public String getQualifiedCatalogName() {
    return null;
  }

  @Override
  public ImmutableMap<String, String> getIssueCodeToLabelMap() {
    return null;
  }

}

```
The classes in the package ddk can be seen in https://github.com/dsldevkit/dsl-devkit/tree/master/com.avaloq.tools.ddk.check.runtime.core/src/com/avaloq/tools/ddk/check/runtime.

**Comment by rubenporras** [comment]
The error is shown as in this screenshot:

![image](https://github.com/eclipse-jdt/eclipse.jdt.core/assets/43636626/b609e4c8-6002-44fe-8cf9-0860d79da730)

**Comment by stephan-herrmann** [comment]
Thanks for the reproducer. By stubbing out all 3rd party classes, I could create a regression test.
The interesting part about this issue: it only affects the editor / reconciler, hence the test is in NullAnnotationModelTests.

Resolved via #1077
